File: UCFPT/21highschoolonline/solutions/radio/radio.java
package solutions.radio;

import java.util.*;
import java.io.*;

public class radio{

	public static void main(String[] args){
		Scanner scan = new Scanner(System.in);
		int t = scan.nextInt();
		for(int q = 1; q <= t; q++){
			int n = scan.nextInt();
			int p = scan.nextInt();
			int s = scan.nextInt();
			String[] freq = new String[n];
			HashMap<String, Integer> map = new HashMap<String, Integer>();
			boolean[] preset = new boolean[n];
			for(int i = 0; i < n; i++){
				freq[i] = scan.next();
				map.put(freq[i], i);
			}
			for(int i = 0; i < p; i++){
				int x = map.get(scan.next());
				preset[x] = true;
			}

			
			// Calculate the distance from each frequency to a preset frequency
			// Cyclic distance can be computed as min(b-a, n-b+a), if a < b
			int[] distFromPreset = new int[n];
			Arrays.fill(distFromPreset, Integer.MAX_VALUE/2);
			for(int i = 0; i < n; i++){
				for(int j = 0; j < n; j++){
					int a = i;
					int b = j;
					if(b < a){
						int temp = a;
						a = b;
						b = temp;
					}
					if(preset[j]) distFromPreset[i] = Math.min(distFromPreset[i], Math.min(b-a, n-b+a));
				}
			}
			
			// For our answer, we either go to the closest preset to the destination frequency
			// or we go directly there. We take the shorter of the two options.
			System.out.println("Day #"+q+":");
			for(int i = 0; i < s; i++){
				int a = map.get(scan.next());
				int b = map.get(scan.next());
				
				int x = distFromPreset[b]+1;
				if(b < a){
					int temp = b;
					b = a;
					a = temp;
				}
				x = Math.min(x, Math.min(b-a, n-b+a));
				System.out.println(x);
			}
			System.out.println();
		}
	}
}

File: UCFPT/21highschoolonline/solutions/radio/radio.cpp
//radio.cpp
//Jacob Steinebronn

#include <bits/stdc++.h>
#define newl '\n'

using namespace std;

int main(){
    int nq;
    cin >> nq;

    for(int dayn=1;dayn<=nq;dayn++){
        cout << "Day #" << dayn << ":\n";
        int n, p, s;
        cin >> n >> p >> s;

        //Store the radio stations
        vector<double> radio(n, 0);

        //Store the indices of the presets
        vector<int> presets(p, -1);


        //Fill the radio[] array with the radio stations
        for(int i=0;i<n;i++)
            cin >> radio[i];
        
        
        //Parse the presets, and fill the presets array with the indices of these stations
        for(int i=0;i<p;i++){
            double pre;
            cin >> pre;
            for(int j=0;j<n;j++){
                if(radio[j]==pre){
                    presets[i]=j;
                    break;
                }
            }
        }

        for(int i=0;i<s;i++){
            double start, finish;
            cin >> start >> finish;
            int idxStart,idxEnd;

            //Find which index we're starting and ending at
            for(int i=0;i<n;i++){
                if(radio[i]==start)
                    idxStart=i;
                else if(radio[i]==finish)
                    idxEnd = i;
            }

            //Without presets, you can either go directly from start to end
            //Or wrap around the end, so take the shorter of these two
            int ans=min(abs(idxStart-idxEnd),n-abs(idxStart-idxEnd));
            
            //For each idx we can get to from taking a preset...
            for(int idx:presets){
                //min ans with 1 + the dist formula from above (the 1 + since it takes one button press to get to the preset)
                ans=min(ans,1+min(abs(idx-idxEnd),n-abs(idx-idxEnd)));
            }

            cout << ans << newl;
        }
        cout << newl;
    }
}
